packages/lib/src/helpers/encodeCall.ts

```diff
@@ -1,15 +1,116 @@
+/*
+ * Utility that wraps around ethereumjs-abi, providing additional validation
+ * on passed type/value pair parameters.
+ */
+
 import abi from 'ethereumjs-abi';
+import * as util from 'ethereumjs-util';
 import BN from 'bignumber.js';
+import _ from 'lodash';
 
-function formatValue(value: any): string {
-  if (typeof(value) === 'number' || BN.isBigNumber(value)) return value.toString();
-  else if (typeof(value) === 'string' && value.match(/\d+(\.\d+)?e(\+)?\d+/)) return (new BN(value)).toString(10);
-  else return value;
-}
+const ERROR_MESSAGE_PREFIX = 'Encoding error';
+const ERROR_MESSAGE = (type: string, value: any) => `${ERROR_MESSAGE_PREFIX} for type ${type} and value ${value.toString()}`;
 
-export default function encodeCall(name: string, args: string[] = [], rawValues: any[] = []): string {
-  const values: string[] = rawValues.map(formatValue);
-  const methodId: string = abi.methodID(name, args).toString('hex');
-  const params: Buffer = abi.rawEncode(args, values).toString('hex');
+export default function encodeCall(name: string, types: string[] = [], values: any[] = []): string {
+  const methodId = abi.methodID(name, types).toString('hex');
+  const params = abi.rawEncode(types, values).toString('hex');
   return '0x' + methodId + params;
 }
+
+export function parseCallValues(types: string[] = [], rawValues: any[] = []): any[] | never {
+  if(types.length !== rawValues.length) throw new Error(ERROR_MESSAGE_PREFIX + '. Supplied number of types and values do not match.');
+  return _.zipWith(types, rawValues, parseTypeValuePair);
+}
+
+export function parseTypeValuePair(type: string, rawValue: any): any | never {
+  if(/^\(.*\)$/.test(type)) return parseTuple(type, rawValue);
+  if(/^[^\[]+\[.*\]$/.test(type)) return parseArray(type, rawValue);
+  if(type === 'address') return parseAddress(type, rawValue);
+  else if(type === 'bool') return parseBool(type, rawValue);
+  else if(type === 'string') return parseString(type, rawValue);
+  else if(type === 'function') return parseFunction(type, rawValue);
+  else if(type.startsWith('bytes')) return parseBytes(type, rawValue);
+  else if(type.startsWith('uint')) return parseNumber(type, rawValue, true, true);
+  else if(type.startsWith('int')) return parseNumber(type, rawValue, false, true);
+  else if(type.startsWith('ufixed')) return parseNumber(type, rawValue, true, false);
+  else if(type.startsWith('fixed')) return parseNumber(type, rawValue, false, false);
+  else throw new Error(ERROR_MESSAGE(type, rawValue) + '. Unsupported or invalid type.');
+}
+
+function parseTuple(type: string, rawValue: any): any | never {
+  if(typeof rawValue === 'string') rawValue = rawValue.split(',');
+  if(rawValue.length === 0) return [];
+  const types = type.replace(/[{()}]/g, '').split(','); // Remove the parenthesis and split the tuple into types.
```

Comment:
Shouldn't we remove only the most outer parentheses? Otherwise, we are losing any kind of tuple nesting, which may lead to errors. For instance:
```
> "(uint,(uint,uint)[])".replace(/[{()}]/g, '')
'uint,uint,uint[]'
```

Reply:
Yes, you're right. I was just removing all parentheses and then splitting by commas, which resulted in a complete flattening of the tuple, which I guess could be ok.
I changed it so that the outer parentheses are removed, and then the string is split by types again (which may include other tuples). Unfortunately, I couldn't come up with a reliable regex for this, so I had to add a quite wordy function `splitTupleTypes(type: string): string[];` :disappointed: 
However, it's consuming super complicated nesting, as you can see in the tests.